Match search input to mocks

mstriemer opened the pull request:
This still needs some cleanup but figured I'd post it for UX comments. @pwalm how does this look?

I'm not crazy about how the cursor appears to the left and then the icon slides over, I'd rather they move together but I couldn't think of a great way to do that.

64px seemed like a lot of padding so I did 16px on each side. Is it supposed to be 64px total? This would be 58px of padding on the input right now (16 \* 2 + 26 [I don't know why the icon is 26px, I guess this should be smaller]).

![search-input-short mov](https://cloud.githubusercontent.com/assets/211578/19662769/9ed10a50-99fe-11e6-8008-4c93eb9ff3fd.gif)

There are 3 elements working together: a full width input that is always left aligned (nothing happens with this so the input works as expected), a label that has the icon and placeholder text, and an icon that does the animation.
- When empty and unfocused the label is shown and centre aligned, JS will position the extra icon underneath the icon you see on load and resize but the extra icon is `opacity: 0` so resizing looks normal.
- When focused the label is hidden using `color: transparent` and `opacity: 0` on its icon, the extra icon immediately gets `opacity: 1` and animates to the left.
- When blurred and there is no text entered the icon animates back, the two icons have matching transitions on opacity with a delay so that once the animation ends the extra icon animates out and the label icon animates in.

tofumatt commented:
I actually like that sliding search box fwiw, and I think it works well.

tofumatt commented on src/ui/components/SearchInput/style.scss:
```diff
@@ -0,0 +1,90 @@
+  color: #ccc;
+  display: inline-block;
+  height: 3rem;
+  line-height: 3rem;
```
I thought the usual rule for line-height was to [just use a number and not a unit](http://meyerweb.com/eric/thoughts/2006/02/08/unitless-line-heights/). (Of course, the label won't have children, right? So it matters less but for consistency's sake I'd like unitless.)

mstriemer replied:
I was setting this to centre the text vertically since the input bar is `3rem`s high. It might not be needed though, I'll at least extract this to a variable if it's still needed.

tofumatt commented on src/ui/components/SearchInput/style.scss:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    [dir=rtl] & {
+      // transform: translateX(-32px);
```
🔥

tofumatt commented on src/ui/components/SearchInput/index.js:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+export default SearchInput;
```
You could move the `export default` above and ditch this line.

tofumatt commented on src/ui/components/SearchInput/index.js:
```diff
@@ -0,0 +1,69 @@
+    const id = `SearchInput-input-${name}`;
+    return (
+      <div className={classNames(className, 'SearchInput')} ref={(el) => { this.root = el; }}>
+        <i className="Icon-magnifying-glass" ref={(el) => { this.icon = el; }} />
```
I know they aren't in place yet but our thoughts of CSS rules (#1056) seem to say we should keep our component name as a common class prefix, so this should be `SearchInput` and not `Icon`.

tofumatt commented on src/ui/components/SearchInput/index.js:
```diff
@@ -0,0 +1,69 @@
+
+  onBlur = () => {
+    if (!this.input.value) {
+      this.root.classList.remove('SearchInput--text');
```
Are we using `--` anywhere else? It seems out of place, though I know there was a thought behind doing it.

mstriemer replied:
I'm not sure if we're using it anywhere else but it feels wrong to call it `SearchInput-text` since there's also a `SearchInput-input`.

I know we had discussed using something like [suitcss's naming conventions](https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md) but we weren't definite on the descendant and utility class name conventions. It would seem I agree with their conventions.

tofumatt commented on src/amo/components/SearchForm.js:
```diff
@@ -33,10 +34,10 @@ export class SearchFormBase extends React.Component {
-               defaultValue={query} className="SearchForm-query" />
+        <SearchInput
+          inputRef={(ref) => { this.searchQuery = ref; }} type="search" name="q"
+          defaultValue={query} className="SearchForm-query" offset={31}
```
What does `offset` do? I couldn't find it elsewhere in the PR.

mstriemer replied:
Leftover from hacks in `CentredInput`. I will remove it.

pwalm commented:
- Can the cursor appear after the slide animation instead of together?
- Should be 64px@xxhdi to the left of the magnifying glass. Might seem big because you're working at xhdi or hdi? (hdi would be 22px)
- Mag icon should be 32px@xxhdi

Looks really good, otherwise. Love the smooth transition.

mstriemer commented:
CSS pixels are at 96DPI then scaled based on screen density and typical viewing distance. Device specific but that's roughly accurate.

I made the icon 16px since I figured it should be half of xxhdi. I wasn't sure of what padding to use so here are some screenshots of different padding amounts. Let me know what works best and I'll go with that.

`padding-start / icon size / padding-end`
<img width="412" alt="screenshot 2016-10-26 09 56 59" src="https://cloud.githubusercontent.com/assets/211578/19731519/4a86ca96-9b63-11e6-8804-7d02701a1f56.png">
<img width="412" alt="screenshot 2016-10-26 09 57 21" src="https://cloud.githubusercontent.com/assets/211578/19731517/4a835b40-9b63-11e6-8e93-89f14c281eaf.png">
<img width="413" alt="screenshot 2016-10-26 09 57 50" src="https://cloud.githubusercontent.com/assets/211578/19731518/4a849b86-9b63-11e6-9c6d-13ebd7611649.png">
<img width="412" alt="screenshot 2016-10-26 09 58 27" src="https://cloud.githubusercontent.com/assets/211578/19731516/4a837dfa-9b63-11e6-8902-38b68da92b12.png">


> Can the cursor appear after the slide animation instead of together?

I have a couple ideas. Neither of them are pretty, but I'll give it a go.

pwalm commented:
24 / 16 / 12 looks the best to me. Thanks for mocking up those options, Mark.

mstriemer commented:
Updated look, the focus is delayed until the transition ends using the lesser of two hacks.

![searchinput mov](https://cloud.githubusercontent.com/assets/211578/19822894/64eea646-9d2b-11e6-94b5-83b60b2c378a.gif)

@tofumatt this should probably get another look at it, I made several changes and I think I addressed all of your comments.

tofumatt commented on src/ui/components/SearchInput/index.js:
```diff
@@ -0,0 +1,88 @@
+  setIconPosition = () => {
+    const { left: labelLeft } = this.labelIcon.getBoundingClientRect();
+    const { left: animateLeft } = this.animateIcon.getBoundingClientRect();
+    this.animateIcon.style.transform = `translateX(${labelLeft - animateLeft}px)`;
```
I forget if I already asked but this is correct in RTL, right?

mstriemer replied:
I believe so but I'll double check. I remember trying to figure out how to handle the RTL case and then I think avoiding it entirely and doing the difference of x positions worked in both cases because `translateX()` is always from the left.

tofumatt commented on src/ui/components/SearchInput/style.scss:
```diff
@@ -0,0 +1,91 @@
+  .SearchInput-animation-icon {
+    color: #fff;
+    opacity: 1;
+    transform: translateX(0) !important;
```
If this `!important` is really needed, a comment explaining why would be handy; it will surely prevent a refactoring future developer from trying to hunt its purpose down trying to remove it. 😉

mstriemer replied:
It's needed because the regular position is set using inline styles. I could remove the value instead but this felt cleaner.

mstriemer commented:
I added a comment that the `!important` is to overwrite inline styles.

RTL works as expected:
![searchinput-rtl mov](https://cloud.githubusercontent.com/assets/211578/19978138/e174a0d4-a1c3-11e6-9056-46524b1d6fbf.gif)

tofumatt commented:
Sweet, looks good to merge!
